[SR-9979] FileHandle class used to implement FileManager._compareFiles

karthikkeyan opened the pull request:
`FileHandler` APIs used to replace the curret implementation of the private method
`_compareFiles(withFileSystemRepresentation:andFileSystemRepresentation:size:bufSize:)`

Bug Link: https://bugs.swift.org/browse/SR-9979

compnerd commented:
@swift-ci please test

spevans commented:
Thanks for this, I just have a few comments:

Since the filenames are already in FSR (File System Representation), converting them back to a `String` is unnecessary, especially since that `String` will then be converted back to an FSR to use in the  `open` call to actually open the file.

I would suggest adding a new `init` method to `FileHandle` along the lines of `internal init?(fileSystemRepresentation: UnsafePointer<Int8>, flags: Int32, createMode: Int)` which takes an FSR instead of a `String`  and defaults to `_closeOnDealloc = true`.

For the read loop, `FileHandle.readData(ofLength:)` is actually quite heavy since it will involve calling `fstat()` and a `malloc()/free()`  for every call in the loop.

I would add another `FileHandle` method, something like:
`internal func _readBytes(into: UnsafePointer<UInt8>, count: Int) throws -> Int`  which just wraps the OS `read` system call and also refactor the inner read loop of `_readDataOfLength` to use it.

This could make the `_compareFiles` look something like:

```
guard let file1 = FileHandle(fileSystemRepresentation: file1Rep, ...) else { return false }
guard let file2 = FileHandle(fileSystemRepresentation: file2Rep, ...) else { return false }

let buffer1 = UnsafeMutablePointer<UInt8>.allocate(capacity: bufSize)
let buffer2 = UnsafeMutablePointer<UInt8>.allocate(capacity: bufSize)
defer {
    buffer1.deallocate()
    buffer2.deallocate()
}


var bytesLeft = size
while bytesLeft > 0 {
    let bytesToRead = Int(min(Int64(bufSize), bytesLeft))
    guard let bytesRead = try? file1._readBytes(into: buffer1, count: bytesToRead),
        bytesRead == bytesToRead else {
	return false
    }
    guard let bytesRead = try? file2._readBytes(into: buffer2, count: bytesToRead),
        bytesRead == bytesToRead else {
	return false
    }
    guard memcmp(buffer1, buffer2, bytesToRead) == 0 else {
	return false
    }
    bytesLeft -= Int64(bytesToRead)
}
    
return true
```

karthikkeyan commented:
@spevans  Thanks for reviewing my PR Simon. I have updated the code as per your suggestion. Now it is ready for the next round of review.

compnerd commented on Foundation/FileHandle.swift:
```diff
@@ -342,6 +342,14 @@ open class FileHandle : NSObject, NSSecureCoding {
     }
+    
+    internal func _readBytes(into buffer: UnsafeMutablePointer<UInt8>, length: Int) throws -> Int {
+        let amtRead = _read(_fd, buffer, length)
```
Hmm, is `_read` the internal method?  Shouldn't that just throw if `read` returned `< 0`?  `_read` is also a UCRT function on Windows, so please use `self._read` to make it more obvious.

karthikkeyan replied:
This `_read` is not seems to be an internal method, it is a reference of either Darwin.read(_:_:_:) or Glibc.read(_:_:_:) method.

karthikkeyan replied:
Since I am do not have experience in Windows, I just want to run by you the following implementation for `_readBytes()` method,

```swift
    internal func _readBytes(into buffer: UnsafeMutablePointer<UInt8>, length: Int) throws -> Int {
#if os(Windows)
        var BytesRead: DWORD = 0
        let BytesToRead: DWORD = DWORD(length)
        if ReadFile(_handle, buffer, BytesToRead, &BytesRead, nil) == FALSE {
            throw _NSErrorWithWindowsError(GetLastError(), reading: true)
        }
        return Int(BytesRead)
#else
        let amtRead = _read(_fd, buffer, length)
        if amtRead < 0 {
            throw _NSErrorWithErrno(errno, reading: true)
        }
        return amtRead
#endif
    }
```

karthikkeyan replied:
If the above code implementation works, I will update the method as well.

compnerd replied:
Yeah, the use of `ReadFile` is fine.

compnerd replied:
Yeah, that implementation is fine.

compnerd commented on Foundation/FileHandle.swift:
```diff
@@ -425,6 +433,15 @@ open class FileHandle : NSObject, NSSecureCoding {
     }
 #endif
+    
+    internal init?(fileSystemRepresentation: UnsafePointer<Int8>, flags: Int32, createMode: Int) {
```
I don't think that we should do this.  The problem is that on Windows, you need to use UTF16 (or Unicode) for the filenames for the FSR otherwise you will not be able to open all files.

karthikkeyan replied:
The methods `_compareFiles()` and `contentEquals()` does not support Windows platform, yet. We will cross the bridge when we get to it. As for as this PR concern it seems to me, it is safe to wrap this `init` method under `#if !os(Windows)`. What do you think @compnerd?

compnerd replied:
The SR was filed specifically to enable Windows support, so I think that it should be part of this change.  Using the `FileHandle` was the way around the FSR conversions.

karthikkeyan replied:
Sounds good. I will update the code to support Windows as well.

karthikkeyan replied:
@compnerd  Thanks for the suggestions. I have updated the init method. Now it supports windows as well. Please do a round of review when you get a chance.

spevans replied:
Maybe a better solution would be to add an internal typealias for FSR
```
#if os(windows)
internal typealias _FileSystemRepresentation = UnsafePointer<UInt16>
#else
internal typealias _FileSystemRepresentation = UnsafePointer<Int8>
#endif
```

And use that in the signatures for these internal functions. I guess `FileManager.fileSystemRepresentation(withPath path: String) -> UnsafePointer<Int8>` is technically broken on Windows anyway but can probably be ignored for now.

compnerd replied:
Hmm, we don't want to break the public interfaces, I think that we should convert across the encodings.

compnerd commented on Foundation/FileManager.swift:
```diff
@@ -1720,39 +1720,35 @@ open class FileManager : NSObject {
             buffer2.deallocate()
         }
-
+        
```
Bleeding whitespace

karthikkeyan replied:
The space between `defer` and the next likes seems necessary, but If you feel strong about this I will update it.

Also while we are in the topic, I have noticed a couple more unnecessary whitespaces as will in the same method; I will remove them too.

spevans replied:
Having an empty line is not an issue, its the trailing whitespace which is the problem (the large green rectangle).

compnerd commented on Foundation/FileHandle.swift:
```diff
@@ -426,6 +434,17 @@ open class FileHandle : NSObject, NSSecureCoding {
+        guard fd > 0 else { return nil }
+        
+#if os(Windows)
+        self.init(handle: HANDLE(bitPattern: _get_osfhandle(fd))!, closeOnDealloc: true)
```
Why can't you use the `init(fileDescriptor:closeOnDealloc:)`?

karthikkeyan replied:
I did not notice that `init` method; I will update the code. Thank you.

karthikkeyan commented:
Thank you @spevans @compnerd, I think I have addressed all the comments, except the new `typealias _FileSystemRepresentation` which I could not quite figure out whats the expectation. Please give it another round of review when you get a chance.

compnerd commented on Foundation/FileHandle.swift:
```diff
@@ -426,6 +443,13 @@ open class FileHandle : NSObject, NSSecureCoding {
+        guard fd > 0 else { return nil }
+        
+        self.init(fileDescriptor: fd, closeOnDealloc: true)
+    }
```
I think that instead of the FSR typealias, we should do something like this:

```
internal convenience init?(fileSystemRepresentation: UnsafePointer<Int8>, flags: Int32, createMode: Int) {
#if os(Windows)
  var fd: Int -1
  if let path = String(cString: fileSystemRepresentation).withCString(encodedAs: UTF16.self) {
    fd = _CFOpenFileWithMode($0, flags, mode_t(createMode))
  }
#else
  let fd = _CFOpenFileWithMode(fileSystemRepresentation, flags, mode_t(createMode))
#endif
  guard fd > 0 else { return nil }
  self.init(fileDescriptor: fd, closeOnDealloc: true)
}
```

karthikkeyan replied:
I think you mean to use `.cString(using: .utf16)` method?

karthikkeyan replied:
It is done.

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

karthikkeyan commented:
@compnerd Looks like the compilation failed. I have built and ran the test cases using Xcode-Beta 10.2 beta 4 (10P107d) using Swift Development Snapshot 2019-03-17 (a) snapshot, successfully. Not sure if it is CI setup issue.

spevans commented:
@swift-ci please test

spevans commented:
@karthikkeyan Don't worry too much about my `typealias _FileSystemRepresentation` idea, it was just a vague notation, I may try prototyping it in a separate PR to see if its any benefit but its not a blocker for this PR.
